Clean up OptimAE

Halbaroth opened the pull request:
This PR refactors the way we manage optimization constraints in
Alt-Ergo.

- Optimization constraints aren't tag on subformulas anymore. Instead
  we use the MaxSMT syntaxes.
- We don't support optimization in Alt-Ergo native format. As this
  feature haven't been published, we don't need to deprecated it.
- We support `maximize`, `minimize` and `get-objectives`. There is no
  support for `assert-soft`. I think it's out of the scope of this PR. 
  Besides, it seems that `get-objectives` can take arguments in our current implementation of `Solving_loop` but I didn't find 
  any examples using this syntax, thus I silently ignore these arguments.
- The SAT solver API exposes two new functions: `optimize` and
  `get_objectives` which, respectively, register objective function and
  return the current objective model.

Halbaroth committed:
Clean up OptimAE

This PR refactors the way we manage optimization constraints in
Alt-Ergo.

- Optimization constraints aren't tag on subformulas anymore. Instead
  we use the MaxSMT syntaxes.
- We don't support optimization in Alt-Ergo native format. As this
  feature haven't been published, we don't need to deprecated it.
- We support `maximize`, `minimize` and `get-objectives`. There is no
  support for `assert-soft`. I think it's out of the scope of this PR.
- The SAT solver API exposes two new functions: `optimize` and
  `get_objectives` which, respectively, register objective function and
  return the current objective model.
- The objective values are expressions.

Halbaroth committed:
Removing the optimization support in the native language

The current implementation of `maximize` and `minimize` isn't satisfying in
the native language. As this feature have never been released and MERCE prefers
using the SMT-LIB format, we remove the support of optimization in the native
language.

Halbaroth committed:
linter

bclement-ocp commented on src/lib/frontend/frontend.ml:
```diff
@@ -429,6 +437,8 @@ module Make(SAT : Sat_solver_sig.S) : S with type sat_env = SAT.t = struct
 
   let th_assume = wrap_f internal_th_assume
 
+  let optimize = handle_sat_exn internal_optimize
```
This needs to be `wrap_f`, not `handle_sat_exn` to avoid using an invalid satML environment (see #946 and related issues).

bclement-ocp commented on src/lib/frontend/models.ml:
```diff
@@ -604,4 +575,3 @@ let output_concrete_model fmt m =
   (*     SmtlibCounterExample.output_arrays_counterexample fmt m.arrays; *)
 
   Printer.print_fmt fmt "@]@,)";
```
```suggestion
  Printer.print_fmt fmt "@]@,)"
```

bclement-ocp commented on src/lib/reasoners/fun_sat.ml:
```diff
@@ -1123,9 +1123,11 @@ module Make (Th : Theory.S) = struct
-        let model = Th.compute_concrete_model env.tbox in
-        env.last_saved_model := model;
-        env
+        match Th.compute_concrete_model env.tbox with
```
When can the theory fail to compute a model?

Halbaroth replied:
I think it cannot happen.

bclement-ocp replied:
Yeah, looks like this returns `None` only when `get_interpretation ()` is disabled (cf other comment)

bclement-ocp commented on src/lib/reasoners/intervalCalculus.ml:
```diff
@@ -2045,17 +2045,35 @@ let case_split env uf ~for_model =
-      | Unknown -> true
-      | _ -> false
-    );
+  let uf, _ = Uf.add uf e in
```
This does not seem right; we don't return the `uf` so the added repr is lost.

Halbaroth replied:
Uhm, I don't think we need to keep the representative element here. The expression `e` here isn't a part of our problem but the expression we try to optimize modulo our constraints. Now, I remember why I add this line before 
```ocaml 
let repr, _ = Uf.find uf e in
```
 
The reason is because the function `Uf.find` fails if the expression have never been added to the union-find before. As the expression can be something completely new, this can happen. Consider for instance the input file:
```
(set-logic ALL)
(set-option :produce-models true)
(declare-const x Int)
(assert (<= x 10))
(maximize (+ x 5))
(check-sat)
(get-model)
(get-objectives)
```
The expression `(+ x 5)` is completely new and the union-find will raise `Not_found` exception here.

Now if the expression is already knew, we want to get both its representative in the union-find (in the code it's `repr`) and the semantic values returned by `X.make` for this expression (in the code it's `r1`).

Another solution would be add the expression `e` in the union-find when we call the function `optimize` of the SAT solver.

bclement-ocp replied:
If we want to get the representative if there is one or the result of `X.make` otherwise then the right code is probably along the lines of:

```ocaml
let repr, r1 =
  match Uf.find uf e with
  | repr, _ -> repr, Uf.make uf e
  | exception Not_found -> let r, _ = X.make e in r, r
```

That said, I didn't think about it, but you are right — we should add `e` in the call to `optimize` (this is what was done before when `Optimize` was a symbol). Note that this needs to use `Theory.add_term` to make sure that all sub-terms are properly added (`Uf.add` only adds the term itself but not the subterms).

The real reason to do this is that if we do `(maximize (bv2nat x))` for instance, and the problem does not otherwise contain `(bv2nat x)`, not adding `bv2nat` to the theory would lose some reasoning power (because it would not be seen by the relations, and knowledge about it would not be properly propagated).

bclement-ocp commented on src/lib/reasoners/intervalCalculus.ml:
```diff
@@ -2045,17 +2045,35 @@ let case_split env uf ~for_model =
 
-let optimizing_split env uf opt_split =
+(* Helper function used in [optimizing_objective] to pick a value
+   for the polynomial [p] in its interval. We used this function in the case
```
```suggestion
   for the polynomial [p] in its interval. We use this function in the case
```

bclement-ocp commented on src/lib/reasoners/intervalCalculus.ml:
```diff
@@ -2045,17 +2045,35 @@ let case_split env uf ~for_model =
   let repr, _ = Uf.find uf e in
   let ty = E.type_info e in
-  let r1 = r in (* instead of repr, which may be a constant *)
+  let r1, _ = X.make e in (* instead of repr, which may be a constant *)
```
Since `e` was added to the union-find this should be `Uf.make uf e` to avoid calling `X.make` twice.

bclement-ocp commented on src/lib/reasoners/sat_solver_sig.ml:
```diff
@@ -97,6 +97,10 @@ module type S = sig
   (* [pred_def env f] assume a new predicate definition [f] in [env]. *)
   val pred_def : t -> Expr.t -> string -> Explanation.t -> Loc.t -> unit
 
+  (* TODO: update this documentation. *)
```
Still a TODO I guess.

bclement-ocp commented on src/lib/reasoners/sat_solver_sig.mli:
```diff
@@ -79,6 +79,10 @@ module type S = sig
   (** [pred_def env f] assume a new predicate definition [f] in [env]. *)
   val pred_def : t -> Expr.t -> string -> Explanation.t -> Loc.t -> unit
 
+  (* TODO: update this documentation. *)
```
Ditto

bclement-ocp commented on src/lib/reasoners/satml.ml:
```diff
@@ -1857,4 +1859,7 @@ module Make (Th : Theory.S) : SAT_ML with type th = Th.t = struct
     enqueue env g.neg 0 None
 
+  let optimize env ~to_max obj =
+    (* TODO: Should we add the objective function to unit_tenv? *)
```
No, the objective function is only relevant when the environment is used for splitting, and the unit environment is not used for splitting.

bclement-ocp commented on src/lib/reasoners/theory.ml:
```diff
@@ -52,16 +52,19 @@ module type S = sig
 
+  val optimize : t -> to_max:bool -> Expr.t -> t
+  (** [optimize env obj] *)
+  (* TODO: update this documentation. *)
```
Still a TODO

bclement-ocp commented on src/lib/util/numbers.mli:
```diff
@@ -155,4 +155,11 @@ module Q : sig
   val sqrt_default : t -> t option
   val sqrt_excess  : t -> t option
+
+  val (~$$) : Z.t -> t
```
Unrelated to the PR but really we should just get rid of the `Numbers` module and use `Z` and `Q` directly, this is a relic of the time we supported either Num or Zarith.

(Well there is the issue of `Q.to_Z` not asserting in Zarith I guess)

Halbaroth replied:
I removed this code and remove the `module Q = numbers.Q` in `intervals`. We'll remove completely the module numbers later.

bclement-ocp replied:
Note, this may be dangerous — I am not sure that some functions such as `Q.floor` or `Q.ceiling` have the same semantics in `Q` and `Numbers.Q`; I'd prefer we did this in a seprate PR.

Halbaroth replied:
... no

bclement-ocp commented on src/lib/structures/typed.mli:
```diff
@@ -285,6 +285,11 @@ and 'a tdecl =
 
+  | TOptimize of Loc.t * 'a atterm * bool
+  (** Optimization declaration.
+      [TOptimize (loc, obj, to_max] declares an objective function [obj]. The
```
```suggestion
      [TOptimize (loc, obj, to_max)] declares an objective function [obj]. The
```

bclement-ocp commented on src/lib/structures/objective.ml:
```diff
@@ -0,0 +1,118 @@
+    to_max : bool;
+  }
+
+  let equal { e = e1; _ } { e = e2; _ } = Expr.equal e1 e2
```
I think this should also compare `to_max`, otherwise if I do `(minimize x)(maximize x)` it has the same effect as `(maximize x)` whereas it should have the same effect as `(minimize x)` (because we must first minimize x *then* maximize it, which does nothing).

Halbaroth replied:
For the input file:
```
(set-logic ALL)
(set-option :produce-models true)
(declare-const x Int)
(assert (<= x 10))
(minimize x)
(maximize x)
(check-sat)
(get-model)
(get-objectives)
```
I got
```
unknown
(
  (define-fun x () Int 0)
)
(objectives 
  (x -oo)
  (x (interval (- oo) (+ oo)))
)
```
It's consistent but the user cannot understand there are not the same objective in the output of `get-objectives`.

bclement-ocp replied:
Ah right, this is handled by the order which will be different.

I think that output is fine, by the way — the user can understand that these are not the same objective because there are two entries, and the goal is that `get-objectives` return one entry per `(minimize)` or `(maximize)` call that was made. And the problem is ill-specified anyways, so. FWIW, Z3 gives a similar output on this problem.

(The intended use case is that there will be a tool that will parse this output and match the list in order with the `minimize` and `maximize` calls)

bclement-ocp commented on src/lib/structures/objective.ml:
```diff
@@ -0,0 +1,118 @@
+  let is_empty = M.is_empty
+  let fold g = M.fold (fun { f; _ } acc -> g f acc)
+
+  let add o v mdl =
```
Nit: I was slightly confused by calling a function `o` (would have expected an order), why not `g`?

Halbaroth replied:
I called it `o` because it's an `objective function`. Maybe `obj`?

Halbaroth commented:
Uhm, I thought it was a good idea to hide completely the priority order in the objective model but in fact, users could feed AE with a file like this one:
```
(set-logic ALL)
(set-option :produce-models true)
(declare-const x Int)
(declare-const y Int)
(assert (<= x 10))
(assert (< y 5))
(maximize x)
(maximize y)
(maximize x)
(check-sat)
(get-model)
(get-objectives)
```
With the current implementation, the second `maximize` will be completely ignored and output:
```
unknown
(
  (define-fun x () Int 10)
  (define-fun y () Real 0.0)
)
(objectives 
  (x 10)
  (y (- 5.0 epsilon))
)
```

Another tricky example is if we use both `maximize` and `minimize` on the same expression. The second one is ignored. Should we print a warning?

bclement-ocp commented on src/lib/reasoners/theory.ml:
```diff
@@ -996,22 +888,35 @@ module Main_Default : S = struct
+    let objectives =
+      Objective.Model.add obj Objective.Value.Unknown env.objectives
+    in
+    let env = reset_case_split_env env in
```
There is a comment in `reset_case_split_env` that states: "be sure it's always correct when this function is called" (wrt the `cs_pending_facts`), but now we can call `reset_case_split_env` from a public function. Are we sure it's always correct?

(I think it is, because from my understanding setting `cs_pending_facts` to `[]` is always correct if we reset `gamma_finite` to `gamma`)

Halbaroth replied:
Actually, I'm not even sure we need to invoke `reset_case_split` here. If it's the first time we register the objective `f`,  we may have propagate case split for `f` because `f` was involved in some assertions. But the objective value of `f` is `Unknown` and we will optimize it anyway.
If we have optimize some subexpressions of `f` here, it means their objectives have a higher priority and there is no reason to throw away these choices.

bclement-ocp replied:
Don't regular (non-optimizing) splits still need to be reset?

For instance if we assert ` 0 ≤ x ≤ 10`, there is currently no optimization under way so we pick `x = 0` as a regular split, then we do `(maximize x)`, we need to undo the `x = 0` choice.

bclement-ocp commented on src/lib/reasoners/theory.ml:
```diff
@@ -996,22 +888,35 @@ module Main_Default : S = struct
-      ~prop_model:assumed_set
-      ~optimized_splits:objectives
-      gamma_finite
+    if Options.get_interpretation () then
```
I don't think this is the right place to have that test. Calling `compute_concrete_model` should always compute a model, but we shouldn't call `compute_concrete_model` if `Options.get_interpretation ()` is false (for similar reasons as #833 ).

Halbaroth replied:
Ok, I will bubble the test.

Halbaroth replied:
I bubble the test in `may_update_last_saved_model`. The current condition to trigger the computation of models is quite complicated. I'm not sure keeping the possibility of generate models at different stages of the SAT solver is very useful.

bclement-ocp commented on src/lib/reasoners/theory.ml:
```diff
@@ -830,14 +722,14 @@ module Main_Default : S = struct
       (* update to optimize with the new gamma *)
-      let objectives, reset_cs_env =
-        update_objectives t.objectives assumed gamma in
+      (* let objectives, reset_cs_env =
+         update_objectives t.objectives assumed gamma in *)
```
Why is this no longer needed? What happens if an assertion renders an existing objective invalid?

Halbaroth replied:
The purpose of `update_objectives` function was to ensure we add new optimization discovered while exploring a term. At this time, we could have optimization everywhere in formulae. In particular, it means the theory module could ignore an optimization constraint before the underlying formula was involved in a formula of the current trail of the SAT solver. I didn't check, but I guess we could even forget an optimization constraint by backtracking.

If we discover a new constraint, we have to throw away all the case splits. A flag in `update_objectives` was used to remember that and we call `reset_case_split_env` if the flag is up.

Now, we feed the SAT solver with all the optimization constraints at once.  I thought it was fine but in fact we perform case splits after each assume. Thus we have to reset casesplits after each `optimize` in `Theory`.

Halbaroth replied:
Actually I did! I just forget to remove the dead code commented. :) Thanks for the remark. I will add a comment in `optimize`.

bclement-ocp replied:
Ah, right, `update_objectives` is for adding new objectives — I mixed it up in my head with `partial_objectives_reset` which is used when an objective becomes invalid, thanks for the clarification!

> I thought it was fine but in fact we perform case splits after each assume.

Yeah, either after each assume or before matching or after matching depending on the split policy. After each assume is the most aggressive version tho, so I think if we deal with that properly we are good.

> I didn't check, but I guess we could even forget an optimization constraint by backtracking.
> Now, we feed the SAT solver with all the optimization constraints at once. I thought it was fine but in fact we perform case splits after each assume. Thus we have to reset casesplits after each optimize in Theory.

This remark makes me wonder: previously, we had optimizations in assertions, which are obviously stored in the assertion stack (and so would be normally affected by `push` and `pop`), but I am not sure that is still the case in this PR (I didn't see any mechanism that would take care of it, but also was not aware it could be an issue prior to your comment). Might be worth checking -- for instance the following should output `0`, not `10`.

```smt-lib
(set-option :produce-models true)
(set-logic ALL)
(declare-const x Int)
(assert (<= 0 x 10))
(push 1)
(maximize x)
(pop 1)
(minimize x)
(check-sat)
(get-model)
```

Halbaroth replied:
Uhm I thought it was fine because I use the same semantic as the assert but I probably miss something in `solving_loop`. Thanks for the test.

bclement-ocp replied:
Ah dang, I take it that means it does not work then :/ I think it is probably an issue in the `Satml_frontend` / `Satml` rather than `solving_loop`. The way `push` and `pop` work in the Satml solver is that we create guards, and assertions within a push/pop assertions are, well, guarded: if the guard is false, the assertion is just ignored.

So for instance the following:

```smt-lib
(push 1)
(assert f1)
(check-sat)
```

is treated as if we had:

```smt-lib
(declare-const guard Bool)
(assert (=> guard f1))
(assert guard) ; Generated for all assertions levels that were not popped on (check-sat)
(check-sat)
```

and the following:

```smt-lib
(push 1)
(assert f1)
(pop 1)
(check-sat)
```

is treated as if we had:

```smt-lib
(declare-const guard Bool)
(assert (=> guard f1))
(assert (not guard)) ; Generated on (pop 1)
(check-sat)
```

and so `Th.assume` (which is called during propagation) will never be called on `f1`. But I believe that with the current implementation (from my recollection of it, I did not check in detail) we will always call `optimize` without taking the guards into account.

In the previous implementation this "just worked" because optimization markers were expressions inside assertions that were not `assume`d when the guards are false, but now that optimization markers are declarations there is a little bit more work that needs to be duplicated. In fact there is probably a bit of overlap with #901, now that I think about it.

Given that optimization is still experimental (and that you've said you're sick of working on it for so long, which is understandable :) ), I think dropping support for optimization within push/pop in this PR is acceptable. We can then either deal with that as part of #901 (which I think is on track for inclusion in the next release, since postponing #904 freed up quite some time for me) or otherwise in a separate PR; your call.

Halbaroth replied:
I think I'll take it after we solve #901. I'll print a warning if we try to optimize with `push` and `pop`. An error isn't necessary here as we still got a correct model but we don't respect the optimization constraint.

Halbaroth replied:
I open an issue #933 and add a warning when we process files involving both optimization constraints and `pop` and `push`.

Halbaroth committed:
improve documentation of objective module

bclement-ocp commented on src/lib/reasoners/theory.ml:
```diff
@@ -552,24 +493,31 @@ module Main_Default : S = struct
     in
-    aux env sem_facts (List.rev env.choices) new_choices
+    aux env sem_facts (List.rev env.choices)
+      (Objective.Model.functions env.objectives, 0) new_choices
```
Is starting at order `0` here really correct? What if some objectives have been optimized already?

bclement-ocp replied:
Hm actually it looks like we are optimizing all objectives again at each case split, that looks wasteful (possibly incorrect, but I don't think so since we use equalities for the splits).

Halbaroth replied:
Let's clarify a bit the previous behaviour and the current one.
- Before this PR, the function `look_for_sat` proceeds choices in this order when `for_model = false`:
  + First, it propagates new choices in the pending queue.
  + Then it try to optimize as much as it can objectives for which we didn't get a best value yet (in the current SAT branch of course). If it see an objective whose the previous value is a limit (+oo, -oo or a strict bound), it tries to optimize it again.
  + If we fail to optimize a bound, we stop in `optimizing_objective`.
  + Otherwise, it asks theories to generate new case splits and it goes back to the first point.

Now, if we're generating model (so `for_model = true`), the previous code ignore objectives whose the value is a limit and we keep optimize next objectives (with a lower priority).

This is not what we want. If we got a limit for some objective `o`, it doesn't make sense to try to optimize objectives with lower priority than `o`. But in model generation, we want to keep generating model, even if we can fulfill some objectives.

Let's pick an example:
``` smt
(set-option :produce-models true)
(set-logic ALL)
(declare-const x Int)
(declare-const y Int)
(maximize x)
(minimize y)
(assert (>= y (- 10)))
(assert (>= x 0))
(check-sat)
(get-model)
(assert (<= x 10))
(check-sat)
(get-model)
```
If you run AE `next` on it with the command
```sh
dune exec -- alt-ergo test.smt2 --optimize --objctives-in-interpretation --frontend legacy --dump-models
```
AE outputs:
```
unknown
Returned unknown reason = Incomplete
(
  (define-fun x () Int +oo)
  (define-fun y () Int (- 10))
)

unknown
Returned unknown reason = Incomplete
(
  (define-fun x () Int 10)
  (define-fun y () Int (- 10))
)
```
We don't expect that AE keep optimizing objectives after `x` before the first `check-sat`, but we keep doing it... 

Now, if we don't the same on this PR with the command
```sh
dune exec -- alt-ergo test2.smt2 --optimize --objectives-in-interpretation
```
AE outputs:
```
unknown
(
  (define-fun x () Int 1)
  (define-fun y () Int 0)
)

; File "test2.smt2", line 12, characters 1-12: I don't know (0.0009) (6 steps) (goal g_2)

unknown
(
  (define-fun x () Int 10)
  (define-fun y () Int (- 10))
)
```

The new code does the right thing... by chance! It's because during each case split phase, we do again all the optimizations, even if we got an infinity value before and we stop optimization if we got an infinity value in `optimizing_objective`. If the flag `for_model` is `true`, we stop also optimization but we keep generate new casesplits. 

I add a new function `next_unknown` in `Objective.Model` whose the semantic is closed to the old `next_optimization`:
```ocaml
  exception Found of Function.t * int

  let next_unknown mdl =
    try
      M.iter (fun { f; order } v ->
        match (v : Value.t) with
        | Value _ -> ()
        | Pinfinity | Minfinity | Limit _ -> raise Exit
        | _ -> raise (Found (f, order))
      ) mdl;
      None
    with
    | Found (f, order) -> Some (f, order)
    | Exit -> None
```

Notice this function isn't very efficient because we keep visiting `Value` node in the map again and again. The behaviour of `Objective.Model` is basically a queue with some printers. I guess I will rewrite it with a functional queue but let's keep map now, early optimization is risky.

Halbaroth replied:
Ok, it's still incorrect for this test:
```smt
; This test checks if the optimization doesn't stop after trying to
; optimize a strict bound.
(set-option :produce-models true)
(set-logic ALL)
(declare-const x Real)
(declare-const y Real)
; If the SAT solver explores first the right branch of the following OR gate, it
; shouldn't conclude that x cannot be optimized.
(assert (or (<= x 4.0) (< x 1.5)))
(maximize x)
; The current implementation of SatML explores the right branch of OR gates
; first. We add the following line to be sure the optimization feature still
; works after changing the order of branches.
(assert (or (< y 1.5) (<= y 4.0)))
(maximize y)
(check-sat)
(get-model)
(get-objectives)
``` 
During the first assert we don't know that `y` will be bound and we get `infinity` here. Thus, we have to treat infinity value again after the second assert, otherwise we will print `+oo` for `y`. This was explained in a comment of the original OptimAE PR.

Gbury replied:
> We don't expect that AE keeps optimize value after `x` after the first `check-sat`, but we keep doing it...

I disagree: in that case I fully expect alt-ergo to re-optimize the model because of the new `get-sat` and the new constraints added by `assert` since the last `check-sat` !

Halbaroth replied:
Ok my example is ambiguous. I was talking of the first `check-sat` only. I totally agree we have to keep optimize if new constraints are involved.

bclement-ocp commented on src/lib/reasoners/theory.ml:
```diff
@@ -474,20 +414,20 @@ module Main_Default : S = struct
            we continue to produce a model if the flag [for_model] is up. *)
         if for_model then
-          aux env sem_facts acc_choices []
+          aux env sem_facts acc_choices ([], 0) []
```
I believe this is a behavior change — previously we would keep optimizing further objectives while doing model generation, but now we no longer does (this call to `aux` goes immediately to `generate_choices`). Is that intentional?

bclement-ocp replied:
Also as far as I can tell, this means that the case splits in the middle of intervals and the such that we create for infinite/limit cases are never used.

bclement-ocp commented on src/lib/reasoners/intervalCalculus.ml:
```diff
@@ -2097,9 +2115,19 @@ let optimizing_split env uf opt_split =
+            Objective.Value.Minfinity
         in
-        Some { opt_split with value }
+        (* As the problem is unbounded, we need to produce a value for the
+           objective function. In this case, we pick a value in the domain
+           of the polynomial [p]. *)
```
Why do we need to explicitly pick a value for the objective? If my understanding is correct, this is only intended to be used during model generation; are the standard model generation mechanism not enough?

Halbaroth replied:
See the comment below for the `Sim.Core.Max`.

bclement-ocp commented on src/lib/reasoners/intervalCalculus.ml:
```diff
@@ -2113,32 +2141,33 @@ let optimizing_split env uf opt_split =
+               in the domain of the polynomial [p]. But we propagate the new
+               value [value] for this objective. Indeed, we want to be able
+               to print it while using the statement [get-objective]. *)
+            middle_value env ~to_max ty p None
```
Same remark as above; are standard model generation mechanisms failing somehow?

Halbaroth replied:
Before this PR, we use a wrapped case-split as input for this function and we have access to the last computed value for the first-order model through this wrapped case-split. 

An alternative implementation would be to give this value as an argument of `optimizing_objective`. 

Notice that my `middle_value` gives a value in the last interval (for the `to_max = true` case) of the domain of the polynomial `p`.  You told me other solvers try to give a bigger value when we maximize strict bound.

bclement-ocp replied:
Ah, right this is for that! Makes sense. What I was thinking for that case (and what other solvers do I believe) is add an inequality constraint rather than an equality in case that specific value end up not being feasible. For instance, with a nonlinear objective `x * x` with `x in `1 .. 2` range, the range for `x * x` is `1 .. 4` (AFAIK Alt-Ergo know this) but if we pick `2` or `3` as a value there is no model whereas if we just constrain `< 3` or `> 2` it is.

(I think in this specific case we might go into an infinite loop in the current PR if we actually used this split in `look_for_sat`, not sure)

Halbaroth replied:
So I suggest to keep the code as it's now. We could make better change later.

bclement-ocp replied:
Sure I am fine with keeping that code as is for now, but the code in `look_for_sat` needs to be fixed to actually use the split during model generation!

bclement-ocp reviewed (commented):
> Uhm, I thought it was a good idea to hide completely the priority order in the objective model but in fact, users could feed AE with a file like this one:
> [..]
> Another tricky example is if we use both maximize and minimize on the same expression. The second one is ignored. Should we print a warning?

Yeah, I was going to make a remark on that. We shouldn't print a warning, we should print all the objectives requested by the user in the correct order!

(We might want to *also* print a warning when we perform optimization under an unbounded/strict objective, which should only occur during model generation if I understand correctly)

Not sure how that's tricky — wouldn't removing the test for an existing function in `Objective.Model.add` just do The Right Thing™?

Stevendeo commented on tests/models/arith/arith10.optimize.smt2:
```diff
@@ -1,8 +1,11 @@
 (set-logic ALL)
+(set-option :produce-models true)
```
This should happen before `(set-logic ...)`

Stevendeo commented on tests/models/arith/arith10.optimize.expected:
```diff
@@ -2,5 +2,9 @@
-  (define-fun y () Real 10.0)
+  (define-fun y () Real 0.0)
 )
+(objectives 
```
As y cannot be optimized, I believe we should warn the user when he invokes (get-objectives)

Stevendeo commented on tests/models/arith/arith4.optimize.expected:
```diff
@@ -4,3 +4,6 @@ unknown
   (define-fun x () Int 1)
   (define-fun y () Int 1)
 )
+(objectives 
```
Why is there only a value for `x`?

Halbaroth replied:
Because we don't try to maximize `y` :D

Stevendeo commented on src/lib/structures/typed.mli:
```diff
@@ -285,6 +285,11 @@ and 'a tdecl =
+  | TOptimize of Loc.t * 'a atterm * bool
+  (** Optimization declaration.
+      [TOptimize (loc, obj, to_max] declares an objective function [obj]. The
+      flag [to_max] determines if we try to maximize of minimize [obj]. *)
```
Don't want to quibble over it, but `to_max` is used as `is_max` everywhere else the code

Halbaroth replied:
Actually I used both. I will put `is_max` everywhere.

Stevendeo commented on src/lib/structures/symbols.ml:
```diff
@@ -155,10 +154,6 @@ let compare_operators op1 op2 =
         let r = Int.compare i1 i2 in
         if r = 0 then Int.compare j1 j2 else r
       | Int2BV n1, Int2BV n2 -> Int.compare n1 n2
-      | Optimize {order=o1; is_max=b1}, Optimize {order=o2; is_max=b2} ->
```
Why is the compare removed?

Stevendeo commented on src/lib/structures/objective.ml:
```diff
@@ -0,0 +1,118 @@
+    | Value w -> Expr.print ppf w
+    | Limit (Above, w) -> Fmt.pf ppf "(+ %a epsilon)" Expr.print w
+    | Limit (Below, w) -> Fmt.pf ppf "(- %a epsilon)" Expr.print w
+    | Unknown -> Fmt.pf ppf "(interval (- oo) (+ oo))"
```
Why the space between `-` (`+`) and `oo` ?

Halbaroth replied:
No good reason. We can consider `oo` as a constant and apply an unary `+` or an unary `-` on it. Or we consider two constants `+oo`, `-oo`. My code is not consistent with that. This is the output of `arith3.optimize.smt2` by `z3`:
```
sat
(
  (define-fun x () Int
    0)
  (define-fun z () Int
    0)
  (define-fun y () Int
    0)
)
(objectives
 (x oo)
 (y  (interval (* (- 1) oo) oo))
 (z  (interval (* (- 1) oo) oo))
)
```
So we can follow the same principal and consider `oo` as a constant and apply usual arithmetical operator on it.

Notice that mathematically, it's a heresy. For instance, `(- oo oo)` doesn't make any sense.

bclement-ocp replied:
FWIW, I believe that `(* (- 1) oo)` is just the way Z3 displays $-\infty$ but that it does not perform other operations on it.

Halbaroth replied:
Sure, I said we have to be consistent and we may use the Z3 convention.

Stevendeo commented on src/lib/structures/objective.ml:
```diff
@@ -0,0 +1,118 @@
+
+  let functions mdl =
+    M.bindings mdl
+    |> List.map (fun ({ f; _ }, _) -> f)
```
There is a way to fold on the map to do this in one go, but I don't expect the map to be too big

Halbaroth replied:
Actually we have to preserve the order of the objectives. If I traverse my map with a fold or a iter, I will get a list in reversed order and I have to iterate on it a second times to get the correct order.

If we use the reversed order in the map, both `add` and `fold` will need to reverse the order :/

bclement-ocp replied:
Given that the current code already iterates on the map twice, folding to build a reversed list then calling `List.rev` is actually going to be slightly faster still, but it is probably not significant.

Stevendeo commented on src/lib/reasoners/satml_frontend.ml:
```diff
@@ -1375,7 +1381,11 @@ module Make (Th : Theory.S) : Sat_solver_sig.S = struct
+
+  let get_model env =
+    Option.bind env.last_saved_model @@
+    fun (lazy mdl) -> Some mdl
```
Isn't that `Option.map Lazy.force env.last_saved_model` ?

Stevendeo commented on src/lib/reasoners/intervals.mli:
```diff
@@ -140,6 +140,10 @@ val add_explanation : t -> Explanation.t -> t
+val pick : to_max:bool -> t -> Numbers.Q.t option
+(** [pick ~to_max t] returns an elements of the set of intervals [t]. If
+    [to_max] is [true], we pick the largest element of [t], if it exists.
+    We look for the smallest element if [to_max] is [false]. *)
```
In the case `t = ]-oo;+oo[`, we don't pick the largest element of `t`

Halbaroth replied:
As I said, `if it exists` :) Otherwise, we pick an element in the last interval of the set of intervals. In the case `]-oo, +oo[`, we pick `0`.

Stevendeo commented on src/lib/reasoners/intervals.ml:
```diff
@@ -1163,11 +1163,41 @@ let match_interval_lower {Sy.sort; is_open; kind; is_lower} i imatch =
+  | ((Strict (q, _) | Large (q, _)), Pinfty) :: _ -> Some Q.(q + ~$1)
+  | ((Strict (q1, _) | Large (q1, _)), (Strict (q2, _) | Large (q2, _))) :: _ ->
+    begin
+      assert (not is_int);
```
This assert is guaranteed by normalization?

Halbaroth replied:
No, it's guaranteed by the pattern matching. All the cases have been matched by a branch before this one if the domain is an integer interval.

Stevendeo commented on src/lib/structures/expr.ml:
```diff
@@ -2682,9 +2683,6 @@ module Purification = struct
         end
 
-      | Sy.Op Sy.Optimize _ ->
-        purify_literal e
```
Is this purification performed somewhere else?

Halbaroth replied:
oups I forget this one. I'll check it now :)

Stevendeo commented on src/lib/reasoners/fun_sat.ml:
```diff
@@ -1849,7 +1851,9 @@ module Make (Th : Theory.S) = struct
   (** returns the latest model stored in the env if any *)
-  let get_model env = !(env.last_saved_model)
+  let get_model env =
+    Option.bind !(env.last_saved_model) @@
```
`Option.map` ?

Halbaroth committed:
change review

Halbaroth committed:
Fix the `look_for_sat` during model generation

Halbaroth committed:
Consistent notation for +oo and -oo

Halbaroth committed:
promote tests

bclement-ocp commented on src/lib/reasoners/satml_frontend.ml:
```diff
@@ -60,12 +62,12 @@ module Make (Th : Theory.S) : Sat_solver_sig.S = struct
     mutable skolems : E.gformula ME.t; (* key <-> f *)
     add_inst : E.t -> bool;
     guards : guards;
+    mutable model_gen_phase : bool;
```
This has been removed in #982

bclement-ocp commented on src/lib/reasoners/satml_frontend.ml:
```diff
@@ -985,11 +988,14 @@ module Make (Th : Theory.S) : Sat_solver_sig.S = struct
       else env.last_saved_model == None
     in
-    if compute then begin
+    if compute && Options.get_interpretation () then begin
```
I believe `compute` being true here already implies `Options.get_interpretation ()` and that this new test is not necessary.

bclement-ocp commented on src/lib/reasoners/satml_frontend.ml:
```diff
@@ -1337,7 +1345,10 @@ module Make (Th : Theory.S) : Sat_solver_sig.S = struct
+  let optimize env ~is_max obj = SAT.optimize env.satml ~is_max obj
+
+  let get_model env =
+    Option.map (fun (lazy mdl) -> mdl) env.last_saved_model
```
```suggestion
    Option.map Lazy.force env.last_saved_model
```

bclement-ocp commented on src/lib/reasoners/intervalCalculus.ml:
```diff
@@ -2097,48 +2115,59 @@ let optimizing_split env uf opt_split =
+        (* As the problem is unbounded, we need to produce a value for the
+           objective function. In this case, we pick a value in the domain
+           of the polynomial [p]. *)
+        let case_split =
+          LR.mkv_eq r1 (middle_value env ~is_max ty p None), true, Th_util.CS
+            (Th_util.Th_arith, Q.one)
```
As far as I can tell, this value is never used ; the `case_split` field is only read when the value is `Value` in `look_for_sat`, never for `Pinfinity` or `Minfinity`. Let's not add dead code.

bclement-ocp commented on src/lib/reasoners/intervalCalculus.ml:
```diff
@@ -2097,48 +2115,59 @@ let optimizing_split env uf opt_split =
+               in the domain of the polynomial [p]. But we propagate the new
+               value [value] for this objective. Indeed, we want to be able
+               to print it while using the statement [get-objective]. *)
+            middle_value env ~is_max ty p None
```
As far as I can tell, this value is never used; the `case_split` field is only read when the value is `Value` in `look_for_sat`, never for `Limit`. Let's not add dead code.

Halbaroth replied:
Now, I propagate it :)

bclement-ocp commented on tests/models/arith/arith7.optimize.expected:
```diff
@@ -3,8 +3,12 @@ unknown
   (define-fun x () Real (/ (- 3) 2))
 )
+(objectives 
+  (x (- (- 0.0 (/ 1 2)) epsilon))
```
This should be displayed as `(- (/ 1 2) epsilon)` following #981 — please rebase the PR.

bclement-ocp commented on src/bin/common/solving_loop.ml:
```diff
@@ -677,35 +682,58 @@ let main () =
+                warning "Some objectives cannot be fulfilled";
+              if State.get suspicious_objectives st then
+                warning "Optimization constraints in presence of push \
+                         and pop statements are not correctly processed.";
```
Nit: it'd be better to flat out reject this, but since it will be fixed ( #993 ) this is OK as well.

bclement-ocp commented on src/bin/common/solving_loop.ml:
```diff
@@ -350,6 +350,10 @@ let main () =
     State.create_key ~pipe:"" "named_terms"
   in
 
+  let suspicious_objectives : bool State.key =
```
This key is not actually related to optimization since what it does is determine whether the problem has `push` or `pop` commands. Its name should reflect that — for instance, `has_push_pop` or `has_incremental_commands`.

bclement-ocp commented on src/bin/common/solving_loop.ml:
```diff
@@ -955,10 +983,16 @@ let main () =
+          match td.contents with
+          | `Pop _ | `Push _ ->
+            State.set suspicious_objectives true st
+          | _ -> st
```
Nit: This is too strict as it prevents any combined use of `(push)` / `(pop)` and optimization, whereas as long as optimization primitives are at the toplevel it should be fine; we could have a cleaner implementation by using an integer key that tracks the current depth and check that we are at level 0 in `handle_optimize_stmt`.

bclement-ocp commented on src/lib/reasoners/fun_sat.ml:
```diff
@@ -1119,12 +1119,12 @@ module Make (Th : Theory.S) = struct
       else !(env.last_saved_model) == None
     in
-    if not compute then env
+    if not compute || not @@ Options.get_interpretation () then env
```
I believe that `compute` can only be true if `Options.get_interpretation ()` is true, and so this check is meaningless.

bclement-ocp commented on src/lib/reasoners/fun_sat.ml:
```diff
@@ -1849,7 +1849,8 @@ module Make (Th : Theory.S) = struct
   (** returns the latest model stored in the env if any *)
-  let get_model env = !(env.last_saved_model)
+  let get_model env =
+    Option.map (fun (lazy mdl) -> mdl) !(env.last_saved_model)
```
```suggestion
    Option.map Lazy.force !(env.last_saved_model)
```

bclement-ocp commented on src/lib/structures/objective.ml:
```diff
@@ -0,0 +1,134 @@
+  let reset_until mdl o =
+    M.fold
+      (fun { f; order } v acc ->
+         if order < o then M.add { f; order } v acc
```
Nit: this could just take `mdl` as the initial value for `acc` and not do anything in this case.

bclement-ocp commented on src/lib/structures/objective.ml:
```diff
@@ -0,0 +1,134 @@
+      M.iter (fun { f; order } v ->
+          match (v : Value.t) with
+          | Value _ -> ()
+          | Limit _ | Pinfinity | Minfinity when for_model -> raise Exit
```
This is a behavior change — the previous code was returning `()` in this case, which allows for further optimization during model generation. I believe the old behavior is better — if we optimize both `x` and `y` and `x` turns out to be unbounded, the user probably would still be happy to have a large value for `y` if possible.

(Recalling that the main use case for optimization is not necessarily to find the exact lexicographic optimum but rather have generally higher-quality models)

Halbaroth replied:
For me it's a very strange behavior but since it's the behavior of some SMT solvers, let's keep the old code.

bclement-ocp reviewed (changes requested):
I think the behavior is mostly correct now — still some minor changes to do, notably I'd like to understand where the case splits for infinities/"limits" are used (or remove them if they are indeed unused).

Halbaroth committed:
rebase artefact

Halbaroth committed:
the lazy force be with you

Halbaroth committed:
Get rid of `0-x`

Halbaroth committed:
rename the key to has_incremental

Halbaroth committed:
keep optimizing after seeing infinity values

Halbaroth committed:
Optimization at toplevel only

The optimization constraint is wrong if we use it in incremental
mode but it's safe if we use `maximize` and `minimize` at the top level
only.

Halbaroth committed:
Remove useless test

Halbaroth committed:
Poetry

Halbaroth committed:
promote tests

Halbaroth committed:
Propagate value produced by `middle_value`

The value produced by `middle_value` in `intervalCalculus` wasn't
propagate to the CC(X) module in `look_for_sat`. We propagate it now.

I also modify the test `arith7.optimize.smt2` as we cannot use both
incremental mode and optimization constraints.

Halbaroth committed:
Prevent impure terms in MaxSMT statements

We don't support yet impure terms or formulae in `maximize` and
`minimize` statements. Let's print a warning message and ignoring
invalid objective functions.

bclement-ocp commented on src/bin/common/solving_loop.ml:
```diff
@@ -350,6 +350,10 @@ let main () =
     State.create_key ~pipe:"" "named_terms"
   in
 
+  let has_incremental : int State.key =
```
If it's an `int`, it should not be called `has_incremental`. I guess two of my comments in the previous round somewhat contradicted each other. Maybe `incremental_depth`?

bclement-ocp commented on src/bin/common/solving_loop.ml:
```diff
@@ -677,35 +682,60 @@ let main () =
+        (State.get solver_ctx_key st).ctx stmt
+    in
+    (* Using both optimization and incremental mode may be wrong if
+       some optimization constraints aren't at the toplevel. *)
```
Should have a link back to #993

bclement-ocp commented on src/lib/reasoners/intervalCalculus.ml:
```diff
@@ -2045,17 +2045,35 @@ let case_split env uf ~for_model =
-      | Unknown -> true
-      | _ -> false
-    );
+  (*   let uf, _ = Uf.add uf e in *)
```
```suggestion
```

bclement-ocp commented on src/lib/reasoners/theory.ml:
```diff
@@ -474,20 +411,22 @@ module Main_Default : S = struct
+          let new_choice =
+            add_explanations_to_split ~order opt_split.case_split
+          in
+          aux env sem_facts acc_choices [new_choice]
```
```suggestion
          propagate_choices env sem_facts acc_choices [new_choice]
```

bclement-ocp commented on src/bin/common/solving_loop.ml:
```diff
@@ -955,10 +985,18 @@ let main () =
+      | td ->
+        let st =
+          match td.contents with
+          | `Pop _ ->
+            State.set has_incremental (State.get has_incremental st - 1) st
+          | `Push _ ->
+            State.set has_incremental (State.get has_incremental st + 1) st
```
This is not +1 and -1; Push and Pop have an integer argument which indicates how many levels are pushed/popped.

```suggestion
          | `Pop n ->
            State.set has_incremental (State.get has_incremental st - n) st
          | `Push n ->
            State.set has_incremental (State.get has_incremental st + n) st
```

bclement-ocp reviewed (approved):
I think we are good with a couple minor omissions noted inline; please fix these before merging!

Halbaroth committed:
Update src/bin/common/solving_loop.ml

Co-authored-by: Basile Clément <129742207+bclement-ocp@users.noreply.github.com>

Halbaroth committed:
Update src/lib/reasoners/intervalCalculus.ml

Co-authored-by: Basile Clément <129742207+bclement-ocp@users.noreply.github.com>

Halbaroth committed:
Update src/lib/reasoners/theory.ml

Co-authored-by: Basile Clément <129742207+bclement-ocp@users.noreply.github.com>

Halbaroth committed:
Link to the issue 993

Halbaroth committed:
rename has_incremental to depth_incremental
